src/drivers/mediatek/mt8186/afe-memif.c

```diff
@@ -0,0 +1,719 @@
+		tr_err(&memif_tr, "MEMIF: Probe failure, unable to allocate channel descriptors");
+		return -ENOMEM;
+	}
+
+	for (channel = 0; channel < dma->plat_data.channels; channel++) {
+		dma->chan[channel].dma = dma;
+		/* TODO need divide to UL and DL for different index */
+		dma->chan[channel].index = channel;
+
+		memif = rzalloc(SOF_MEM_ZONE_SYS_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+				sizeof(struct afe_memif_dma));
+		if (!memif) {
+			tr_err(&memif_tr, "afe-memif: %d channel %d private data alloc failed",
+			       dma->plat_data.id, channel);
+			goto out;
+		}
+
+		memif->afe = afe;
+		memif->memif_id = channel;
+		dma_chan_set_data(&dma->chan[channel], memif);
+	}
+	return 0;
+
+out:
+	if (dma->chan) {
+		for (channel = 0; channel < dma->plat_data.channels; channel++)
+			rfree(dma_chan_get_data(&dma->chan[channel]));
+		rfree(dma->chan);
+		dma->chan = NULL;
+	}
+
+	afe_remove(afe);
+
+	return -ENOMEM;
+}
+
+static int memif_remove(struct dma *dma)
+{
+	int channel;
+	struct mtk_base_afe *afe = afe_get();
+
+	if (!dma->chan) {
+		tr_err(&memif_tr, "MEMIF: remove called without probe, it's a no-op");
+		return 0;
+	}
+	for (channel = 0; channel < dma->plat_data.channels; channel++) {
+		/* TODO Disable HW requests for this channel */
+
+		rfree(dma_chan_get_data(&dma->chan[channel]));
+	}
+	rfree(dma->chan);
+	dma->chan = NULL;
+
+	return afe_remove(afe);
+}
+
+static int memif_interrupt(struct dma_chan_data *channel, enum dma_irq_cmd cmd)
+{
+	int ret;
+	unsigned int period;
+	unsigned int sample_size;
+	struct afe_memif_dma *memif;
+	struct mtk_base_afe *afe = afe_get();
+
+	if (channel->status == COMP_STATE_INIT)
+		return 0;
+
+	memif = dma_chan_get_data(channel);
+	assert(memif);
+
+	switch (cmd) {
+	case DMA_IRQ_STATUS_GET:
+		ret = afe_irq_get_status(afe, memif->irq_id);
+		break;
+	case DMA_IRQ_CLEAR:
+		ret = afe_irq_clear(afe, memif->irq_id);
+		break;
+	case DMA_IRQ_MASK:
+		ret = afe_irq_disable(afe, memif->irq_id);
+		break;
+	case DMA_IRQ_UNMASK:
+		sample_size = ((memif->format == SOF_IPC_FRAME_S16_LE) ? 2 : 4) * memif->channel;
+		period = memif->period_size / sample_size;
+		ret = afe_irq_config(afe, memif->irq_id, memif->rate, period);
+		if (ret < 0)
+			return ret;
+		ret = afe_irq_enable(afe, memif->irq_id);
+		if (ret < 0)
+			return ret;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return ret;
+}
+
+/* TODO need convert number to platform MACRO */
+static int memif_get_attribute(struct dma *dma, uint32_t type, uint32_t *value)
+{
+	switch (type) {
+	case DMA_ATTR_BUFFER_ALIGNMENT:
+	case DMA_ATTR_COPY_ALIGNMENT:
+		*value = 4;
+		break;
+	case DMA_ATTR_BUFFER_ADDRESS_ALIGNMENT:
+		*value = 16;
+		break;
+	case DMA_ATTR_BUFFER_PERIOD_COUNT:
+		*value = 4;
+		break;
+	default:
+		return -ENOENT;
+	}
+	return 0;
+}
+
+static int memif_get_data_size(struct dma_chan_data *channel, uint32_t *avail, uint32_t *free)
+{
+	uint32_t hw_ptr;
+	struct afe_memif_dma *memif = dma_chan_get_data(channel);
+
+	assert(memif);
```

Comment:
avoid asserts if possible please, we are trying to remove them

Reply:
thanks, I'll remove them